gms/gossiper.cc

```diff
@@ -224,14 +224,58 @@ future<> gossiper::handle_syn_msg(msg_addr from, gossip_digest_syn syn_msg) {
         return make_ready_future<>();
     }
 
-    auto g_digest_list = syn_msg.get_gossip_digests();
-    do_sort(g_digest_list);
-    utils::chunked_vector<gossip_digest> delta_gossip_digest_list;
-    std::map<inet_address, endpoint_state> delta_ep_state_map;
-    this->examine_gossiper(g_digest_list, delta_gossip_digest_list, delta_ep_state_map);
-    gms::gossip_digest_ack ack_msg(std::move(delta_gossip_digest_list), std::move(delta_ep_state_map));
-    logger.trace("handle_syn_msg(): response={}", ack_msg);
-    return this->ms().send_gossip_digest_ack(from, std::move(ack_msg));
+    syn_msg_pending& p = _syn_handlers[from.addr];
+    if (p.pending) {
+        // The latest syn message from peer has the latest infomation, so
+        // it is safe to drop the previous syn message and keep the latest
+        // one only.
+        logger.debug("Queue gossip syn msg for node {}, syn_msg={}", from, syn_msg);
+        p.syn_msg = std::move(syn_msg);
+        return make_ready_future<>();
+    } else {
+        // Process the syn message immediately
+        logger.debug("Process gossip syn msg for node {}, syn_msg={}", from, syn_msg);
+        return do_send_ack_msg(from, std::move(syn_msg));
```

Comment:
I think it would be simpler to maintain _syn_handlers here (with one futurize_apply and finally), and have do_send_ack_msg() not deal with it at all, only deal with handling. 

If you also embrace the idea to simply ignore duplicated SYNs, then it's really simple with `gate<>`:


```
if (syn_msg_pending.active_acks.get_count()) {
  // ignore duplicated SYN
  return make_ready_future<>();
}
return with_gate(syn_msg_pending.active_acks, [] {
   return do_send_ack_msg(from, std::move(syn_msg));
});
```

Reply:
This comment was not addressed (the first part, which is about making do_send_ack_msg not deal with _syn_handlers at all). Same about the new `do_send_ack2_msg()`